File: deploy/pull-db.pl
#!/usr/bin/env perl

=head1 SYNOPSIS

    pull-db.pl [--production]
    options:
      --production      Pull from questhub.io instead of dropbox backup
      --keep-settings   Don't clean up email and other sensitive settings

=cut

use strict;
use warnings;

use 5.010;

use Getopt::Long 2.33;
use Pod::Usage;

use IPC::System::Simple;
use autodie qw(system);

sub docker_do {
    my ($container, @args) = @_;
    system('docker', 'exec', '-it', "questhub_${container}_1", @args);
}

sub mongo_do {
    docker_do('mongo', @_);
}

sub backend_do {
    docker_do('backend', @_);
}

sub main {
    my $from_production;
    my $keep_settings;
    GetOptions(
        'p|production!' => \$from_production,
        'k|keep-settings!' => \$keep_settings,
    ) or pod2usage(2);
    pod2usage(2) if @ARGV;

    my $HOST = 'questhub.io';

    backend_do('./clear_mongo.sh');
    mongo_do('mongo', 'play', '--eval', 'db.realms.drop()');

    system('rm -rf dump');
    system('mkdir dump');

    if ($from_production) {
        system(qq{ssh ubuntu\@$HOST "sh -c 'rm -rf dump && mongodump -d play'"});
        system(qq{scp -r ubuntu\@$HOST:dump/play ./dump/play});
    }
    else {
        my $backup_file = qx(ls -t1 ~/Dropbox/backup/$HOST/ | head -1);
        unless (defined $backup_file) {
            die "backup file not found: $!";
        }
        chomp $backup_file;

        system("tar xfvz ~/Dropbox/backup/$HOST/$backup_file");
    }

    system("tar -c dump | docker exec -i questhub_mongo_1 bash -c 'cd /data && rm -rf dump && tar -xv && cd dump && mongorestore --drop .'");

    # to avoid accidentally sending emails to users while debugging
    mongo_do('mongo', 'play', '--eval', 'db.users.update({}, {"$unset": { "settings" : 1 } }, false, true)') unless $keep_settings;
}

main unless caller;

File: deploy/pull-db.py
#!/usr/bin/env python3

import argparse
import subprocess

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--production', action='store_true')
    parser.add_argument('--keep-settings', action='store_true')

    args = parser.parse_args()

    HOST = 'questhub.io'

    subprocess.check_call('docker exec -it questhub_backend_1 ./script/clear_mongo.sh', shell=True)
    subprocess.check_call('docker exec -it questhub_mongo_1 mongo play --eval "db.realms.drop()"', shell=True)
    subprocess.check_call('rm -rf dump', shell=True)
    subprocess.check_call('mkdir dump', shell=True)

    if args.production:
        subprocess.check_call([
            'ssh', 'ubuntu@{}'.format(HOST),
            "sh -c 'rm -rf dump && mongodump -d play'"
        ])
        subprocess.check_call([
            'scp', '-r', 'ubuntu@{}:dump/play'.format(HOST),
            './dump/play'
        ])
    else:
        backup_file = subprocess.check_output(
            'ls -t1 ~/Dropbox/backup/{} | head -1'.format(HOST),
            shell=True,
            universal_newlines=True,
        ).strip()
        subprocess.check_call('tar xfvz ~/dropbox/backup/{}/{}'.format(HOST, backup_file), shell=True)

    subprocess.check_call("tar -c dump | docker exec -i questhub_mongo_1 bash -c 'cd /data && rm -rf dump && tar -xv && cd dump && mongorestore --drop .'", shell=True)

    if not args.keep_settings:
        subprocess.check_call([
            'docker', 'exec', '-it', 'questhub_mongo_1',
            'mongo', 'play', '--eval', 'db.users.update({}, {"$unset": { "settings": 1 }}, false, true)'
        ])


if __name__ == '__main__':
    main()
